api/envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -0,0 +1,142 @@
+
+ // Required. List of principals that are assigned/denied the role based on “action_type”.
+ repeated Principal principals = 2 [(validate.rules).repeated .min_items = 1];
+}
+
+message StringMatch {
+  oneof match_pattern {
+    // The input string must match exactly the string specified here.
+    // Or it is a "*", which means that it matches any string.
+    string simple = 1;
+
+    // The input string must have the prefix specified here.
+    string prefix = 2;
+
+    // The input string must have the suffix specified here.
+    string suffix = 3;
+
+    // The input string must match the regular expression specified here.
+    // The regex grammar is defined `here
+    // <http://en.cppreference.com/w/cpp/regex/ecmascript>`_.
+    string regex = 4;
+ }
+}
+
+// Specifies how to match an entry in a map.
+message MapEntryMatch {
+  // The key to select an entry from the map.
+  string key = 1;
+
+  // A list of matched values.
+  repeated StringMatch values = 2;
+}
+
+// Permission defines a permission to access the service.
+message Permission {
+  // Optional. A list of HTTP paths or gRPC methods.
+  // gRPC methods must be presented as fully-qualified name in the form of
+  // packageName.serviceName/methodName.
+  // If this field is unset, it applies to any path.
+  repeated StringMatch paths = 1;
+
+  // Required. A list of HTTP methods (e.g., "GET", "POST").
+  // If set to ["*"], it applies to any method.
+  repeated string methods = 2;
+
+  // Definition of a custom condition.
+  message Condition {
+    oneof condition_spec {
+      // Header match. This matches to the "request.http.headers" field in
+      // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+      // The map key is the header name. The header specifies how the service is accessed.
+      MapEntryMatch header = 1;
+    }
+  }
+
+  // Optional. Custom conditions.
+  repeated Condition conditions = 3;
+}
+
+// Principal defines an identity or a group of identities.
+message Principal {
+  // Optional. The name of the principal. This matches to the "source.principal" field in
+  // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+  // If set to "*", it applies to any user.
+  string name = 1;
+
+  // Definition of a custom attribute to identify the principal.
+  message Attribute {
+    oneof attribute_spec {
+      // Source service name. This matches to the "source.service" field in
+      // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+      string service = 1;
+
+      // Source IP addresses.
+      repeated envoy.api.v2.core.CidrRange ips = 2;
+
+      // Header match. This matches to the "request.http.headers" field in
+      // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
+      // The map "key" is the header name. The header identifies the client.
+      MapEntryMatch header = 3;
```

Comment:
I get your point that headers that are used for permissions and for principal constraints might be different, but the flip side is that I'm not sure how they will be treated any differently by the proxy code. I guess I have a larger question about how processing occurs for principal constraints that I think it's important that we clarify now.

Based on our conversations, my understanding is that we would like another filter(s) to set principal prior to this filter being invoked. Are the previous filter(s) supposed to operate on these constraints? Or does the RBAC filter use the constraints as a "post-filter?"

If the latter, does it allow me to have multiple principal policies in which the principal is '*' but then I select the actual policy based on the principal constraints? If so, is that in FIFO matching order?

Also, if the latter, it seems like authn filters are not going to expose the entire Principal object, but instead just the name? If that's the case, I would suggest renaming `Principal` as you define it to something else (`PrincipalMatch` or something) and actually putting `name` into a new message called `Principal` that we can assume will be set by some other filter. The lack of existence of `Principal` can replace '*' to mean match any principal? I think that is better?

Reply:
Having an authn filter in front of RBAC filter is preferred (so that we have authenticated attributes), but is NOT required. This is what I have in mind. 
* When there is a RBAC configuration change, the RBAC policies are passed to the RBAC engine, which is a C++ library. 

* At runtime when a request arrives, 
1. We will have some code (in RBAC filter, or in authn filter before RBAC) that fill out the AttributeContext. This is similar to the external authz implementation [here](https://github.com/envoyproxy/envoy/blob/71df74ad6680629d0c31a2328616dd0d5dae4c4d/source/extensions/filters/common/ext_authz/ext_authz_impl.cc#L179). 
2. The AttributeContext (request context) is passed to RBAC engine. The RBAC engine evaluates the AttributeContext against the RBAC policy. 

The logic in RBAC engine can be (suppose action_type is ALLOW):
```
* Go through each policy, check if 
  * the action is allowed by one of the "permissions", AND
  * the caller matches one of the "principals". 
  If so, allow the request. 

The request is allowed if any policy allows it. A request is denied if no policy allows it. 
```

Hence, if the principal name is "*", all users will have the given permission. The order of the policy does not matter. If you would like to put some constraints on the principal, do not use name: "*", but directly specify the constraints. For example, the following spec specifies a principal that has two constraints (1) source IPs is in 208.130.28.0/22 (2) "x-api-key" header is "mykey". 

```
principals:
- attributes:
  - source_ips: 
      cidrs: 
      - address_prefix: 208.130.28.0
        prefix_len: 22
  - header:
      key: "x-api-key"
      values: ["mykey"]
```

Reply:
cc. @rodaine

Reply:
> When there is a RBAC configuration change, the RBAC policies are passed to the RBAC engine, which is a C++ library.

The RBAC engine is the filter. Just to keep using the same language let's just assume that for now.

OK I think I understand. So basically, you are assuming that "name" is the only thing that is set by the authentication filter, and that effectively name is optional? If so, that makes sense to me, but here are the changes that I would make. 

1) Rename `Principal` to `PrincipalMatch`. 
2) Make a new message `Principal` that contains a single string field `name`. The object `Principal` is what can be set by optional authn filters.
3) Instead of "*", just use the presence or absence of the `Principal` object to denote whether to match on the principal name (or other fields in the future).

I think if we make the above changes I'm happy with this solution and we can move forward. WDYT?

Reply:
Thanks @mattklein123. Yes, "name" in Principal is optional. 

Just to clarify, do you mean separating the authentication principal fields vs. un-authenticated principal fields in different messages? We may have the situation that the same field (e.g., service) which is currently unauthenticated, but may later be added to JWT claim or Cert, and hence becomes authenticated. From RBAC engine point of view, it simply matches policies with request context (AttributeContext). Populating the AttributeContext can be done either by an authn filter, or in RBAC filter itself (before invoking Check(AttributeContext)), or in both. 

"principals" and "permissions" comprise a "policy", which defines "who" (principal) can "do what" (permission). The "xxxMatch" (StringMatch, MapEntryMatch, IpMatch, PortMatch) are for matching a particular field (i.e., attribute) of that type (string type, map type, etc). Calling the current "Principal" message "PrincipalMatch" does not fit well with the current categorization, because principal is not an attribute, it is composed of a set of attributes that describe the caller ("who").

Reply:
> Just to clarify, do you mean separating the authentication principal fields vs. un-authenticated principal fields in different messages?

Yes. What I'm really asking is: If someone uses a different authn filter to set principal, what is that filter setting exactly? I don't understand how that filter would be setting things that we might want to match on. This is why I'm suggesting splitting Principal (which authn filter might set) from the authz matching infrastructure which is what you have here. Call things whatever you like beyond that...

Reply:
I added "Authenticated" message in Principal. So principal has two parts, authenticated attributes, and custom attributes. Currently authenticated attributes only include the "name", in the future, we can add, say, JWT claims.

Reply:
Thanks @liminw. This last change is great. I have one remaining clarification that probably just needs a doc update/clarification. 

Let's say I want to have different permissions for principals with different attributes. Specifically, let's say I define something like:

- principal1: x-some-header: 4, source_ips: 1.2.3.4/32
- principal2: x-some-header: 5, source_ips: 1.2.3.4/24

How do I select which principal is used for my decision? Do I iterate through the list in order? What if there are multiple overlapping matches? Is it like routing where first wins?

Reply:
OK, great, that's what I would expect also. Can you add some clarifying docs/examples on this type of matching, and then I think we are good to go?

Reply:
I added a comment at the beginning of rbac.proto. Thanks!